state/relationunit.go

```diff
@@ -356,19 +356,42 @@ func (ru *RelationUnit) Valid() (bool, error) {
 	if ru.endpoint.Scope != charm.ScopeContainer || ru.isPrincipal {
 		return true, nil
 	}
-	// A subordinate container-scoped relation unit is only valid if
-	// its principal unit is also a member of the relation.
+	// A subordinate container-scoped relation unit is valid if:
+	// the other end of the relation is also a subordinate charm
+	// or its principal unit is also a member of the relation.
+	appName, err := names.UnitApplication(ru.unitName)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	var otherAppName string
+	for _, ep := range ru.relation.Endpoints() {
+		if ep.ApplicationName != appName {
+			otherAppName = ep.ApplicationName
+		}
+	}
+	if otherAppName == "" {
+		return false, errors.Errorf("couldn't find other endpoint")
+	}
+	app, err := ru.st.Application(otherAppName)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	if !app.IsPrincipal() {
+		return true, nil
+	}
+
 	unit, err := ru.st.Unit(ru.unitName)
 	if err != nil {
 		return false, errors.Trace(err)
 	}
 	// No need to check the flag here - we know we're subordinate.
-	name, _ := unit.PrincipalName()
-	appName, err := names.UnitApplication(name)
+	pName, _ := unit.PrincipalName()
+	principalAppName, err := names.UnitApplication(pName)
 	if err != nil {
 		return false, errors.Trace(err)
 	}
-	_, err = ru.relation.Endpoint(appName)
+	// If the other application is a principal, only allow it if it's in the relation.
+	_, err = ru.relation.Endpoint(principalAppName)
 	return err == nil, nil
```

Comment:
Yeah, I'll tighten that up - at the moment the only error `Endpoint` can return is the one indicating that the app passed in isn't an endpoint, but this code shouldn't assume that.